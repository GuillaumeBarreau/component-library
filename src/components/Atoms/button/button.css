/* Button */
.button {
   background-color: #f0eeee;
   border: none;
   cursor: pointer;
   outline: none;
   border-radius: 1.25rem;
   margin: 0.5rem;
   padding: 0.25rem 1.5rem;
   justify-content: center;
   border-radius: 12px;
   display: flex;
   flex-direction: row;
   font-family: Black;
   text-transform: uppercase;
   transition: 0.2s;
}

.button::first-letter {
   text-transform: capitalize;
}

.button-default {
   border-color:#24262e;
   color: var(--color-base-01);
   box-shadow: 0px 6px 0px 0px var(--color-base-02--dark);
}

.button-default:active {
   background: var(--color-base-02--dark);
}

.button-default--variant-outlaned {
   border: 2px solid var(--color-base-02--dark);
}

.button-default--variant-contained {
   background: var(--color-notice-error);
   color: white;
}

.button-default--variant-contained {
   background: Black;
}

.button-error {
   border-color: var(--color-notice-error);
   color: var(--color-notice-error);
   box-shadow: 0px 6px 0px 0px var(--color-notice-error);
}

.button-error:active {
   background: var(--color-notice-error);
}

.button-error--variant-outlaned {
   border: 2px solid var(--color-notice-error);
}

.button-error--variant-contained {
   background: var(--color-notice-error);
   color: white;
}

.button-success {
   border-color: var(--color-notice-success);
   color: var(--color-notice-success);
   box-shadow: 0px 6px 0px 0px var(--color-notice-success);
}

.button-success:active {
   background: var(--color-notice-success);
}

.button-success--variant-outlaned {
   border: 2px solid var(--color-notice-success);
}

.button-success--variant-contained {
   background: var(--color-notice-success);
   color: white;
}

.button-warning {
   border-color: var(--color-notice-warning);
   color: var(--color-notice-warning);
   box-shadow: 0px 6px 0px 0px var(--color-notice-warning);
}

.button-warning:active {
   background: var(--color-notice-warning);
}

.button-warning--variant-outlaned {
   border: 2px solid var(--color-notice-warning);
}

.button-warning--variant-contained {
   background: var(--color-notice-warning);
   color: white;
}

.button-info {
   border-color: var(--color-notice-info);
   color: var(--color-notice-info);
   box-shadow: 0px 6px 0px 0px var(--color-notice-info);
}

.button-info:active {
   background: var(--color-notice-info);
}

.button-info--variant-outlaned {
   border: 2px solid var(--color-notice-info);
}

.button-info--variant-contained {
   background: var(--color-notice-info);
   color: white;
}

.button-info:active,
.button-error:active,
.button-warning:active,
.button-success:active,
.button-default:active {
   box-shadow: none;
   transform: translateY(6px);
   transition: 0.3s;
   color: var(--color-base-02--light);
}

.button-disabled {
   color: #646060;
   opacity: 0.8;
   cursor: not-allowed;
}

.button-disabled--variant-outlaned {
   border: 2px solid #646060;
}

[class*="--variant-outlaned"],
[class*="--variant-contained"] {
   box-shadow: none;
   background-color: none;
}