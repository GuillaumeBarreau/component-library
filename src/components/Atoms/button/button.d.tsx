export interface ComponentButtonProps {
  className?: string;
  handleClick: React.MouseEventHandler<HTMLButtonElement>;
  iconName?: string;
  type?: string;
  variant?: "contained" | "outlaned"  | "default"
  notice?:
    | "success"
    | "info"
    | "warning"
    | "error"
    | "disabled"
    | "default"
    | undefined;
  children: string;
}
