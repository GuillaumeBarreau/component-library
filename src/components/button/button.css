/* Button */
.button {
   background-color: #f9f9f9;
   border: none;
   cursor: pointer;
   outline: none;
   border-radius: 1.25rem;
   margin: 0.5rem;
   padding: 0.25rem 1.5rem;
   justify-content: center;
   border-radius: 12px;
   display: flex;
   flex-direction: row;
   font-family: Black;
   text-transform: uppercase;
   transition: 0.2s;
}

.button::first-letter {
   text-transform: capitalize;
}

.button-default {
   border-color:#24262e;
   color: var(--color-base-01);
   box-shadow: 0px 6px 0px 0px var(--color-base-02--dark);
}

.button-default:active {
   background: var(--color-base-02--dark);
   color: #0b0b0e;
}

.button-error {
   border-color: var(--color-notice-error);
   color: var(--color-notice-error);
   box-shadow: 0px 6px 0px 0px var(--color-notice-error);
}

.button-error:active {
   background: var(--color-notice-error);
}

.button-success {
   border-color: var(--color-notice-success);
   color: var(--color-notice-success);
   box-shadow: 0px 6px 0px 0px var(--color-notice-success);
}

.button-success:active {
   background: var(--color-notice-success);
}

.button-warning {
   border-color: var(--color-notice-warning);
   color: var(--color-notice-warning);
   box-shadow: 0px 6px 0px 0px var(--color-notice-warning);
}

.button-warning:active {
   background: var(--color-notice-warning);
}

.button-info {
   border-color: var(--color-notice-info);
   color: var(--color-notice-info);
   box-shadow: 0px 6px 0px 0px var(--color-notice-info);
}

.button-info:active {
   background: var(--color-notice-info);
}

.button-info:active,
.button-error:active,
.button-warning:active,
.button-success:active,
.button-default:active {
   box-shadow: none;
   transform: translateY(6px);
   transition: 0.3s;
   color: var(--color-base-02--light);
}

.button-disabled {
   color: #646060;
   opacity: 0.8;
   cursor: not-allowed;
}

