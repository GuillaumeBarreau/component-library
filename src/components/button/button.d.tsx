
export interface ComponentButtonProps {
  className?: string;
  handleClick: React.MouseEventHandler<HTMLButtonElement>;
  iconName?: string;
  type?: string;
  notice?: "success" | "info" | "warning" | "error" | "disabled" | "default" | undefined;
  children: string;
}
