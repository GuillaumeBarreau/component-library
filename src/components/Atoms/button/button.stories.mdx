import { Meta, Story, Canvas, ArgsTable } from "@storybook/addon-docs";
import { Button } from "./button";

<Meta 
  title="Components/Atoms/Button" 
  component={Button}
  argTypes={{
    notice: {
      control: "select",
      description: "Modify the button color or used disable when the user cannot proceed or until an input is collected.",
      options: [
        "default",
        "info",
        "warning",
        "success",
        "error",
        "disabled",
        "primary",
        "secondary",
      ],
    },
    className: {
      description: "Override a component's styles",
      control: "text",
    },
    variant: {
      control: "select",
      options: ["text", "outlined"],
    },
    iconName: {
      control: "select",
      description: "Icon displayed on the left of the button",
      options: [""],
    },
    isBlock: {
      description: "take the full width of its parent element",
      control: "boolean",
    },
    handleClick: {
      description: "Function called when the user clicks on the button",
      control: "boolean",
    },
    children: {
      control: "text",
      description: "Children of the button.",
    }
  }}
  args={{
    children: 'children',
    handleClick: () => console.log("test")
  }}/>

# Button

The Button component is used to trigger an action or event, such as submitting a form,
opening a Dialog, canceling an action, or performing a delete operation.

## React Component

#### Import library

```
Import {Atoms} from '@bonjour-les-pixels/component-library';
```

#### Use Buttom component

```
<Atoms.Button {...props}> CHILDREN_TEXT </Atoms.Button>
```

## Playground

<Canvas>
  <Story name="Button">
    {args => {
      return <Button {...args}  />;
    }}
  </Story>
</Canvas>

### Props

<ArgsTable story="Button" />

### Notice

You can use the `notice` prop to modify the button color. This props can be declined in 7 ways. The style varies according to the level of importance of the action or the brand color in the page context.

<Canvas>
    <Story name="Button/Notice">
        <Button
          notice="primary"
          handleClick={() => console.log("here")}
          size="medium"
          children="Primary"
        />
        <Button
          notice="secondary"
          handleClick={() => console.log("here")}
          size="medium"
          children="secondary"
        />
        <Button
          handleClick={() => console.log("here")}
          size="medium"
          children="Default"
        />
        <Button
          notice="success"
          handleClick={() => console.log("here")}
          size="medium"
          children="success"
        />
        <Button
          notice="info"
          handleClick={() => console.log("here")}
          size="medium"
          children="info"
        />
        <Button
          notice="error"
          handleClick={() => console.log("here")}
          size="medium"
          children="error"
        />
        <Button
          notice="warning"
          handleClick={() => console.log("here")}
          size="medium"
          children="warning"
        />
     </Story>
</Canvas>

You can use the value `disabled` in `notice` prop when you need to keep a user from using the button component until some specific condition has been met.

<Canvas>
  <Button
      notice="disabled"
      handleClick={() => null}
      children="disabled"
    />
</Canvas>

### ClassName

You can use the `className` prop to override a component's styles using custom classes.

```
  <Button
  className="override-style-componenent"
/>
```

### HandleClick

You can use the `HandleClick` prop for determine what action is to be taken whenever an event is trigger.
This prop is required.

```
  <Button
  HandleClick={() => alert('event trigger')}
/>
```

<Canvas>
  <Story name="Button/HandleClick">
    <Button
        handleClick={() => alert('event trigger')}
        children="handleClick"
      />
  </Story>
</Canvas>

### IconName

You can use the iconName prop to have icons. You will find a list icons available below. 

<Canvas>
  <Story name="Button/IconName">
    <Button
      handleClick={() => console.log("here")}
      iconName="coffee"
      children="button"
    />
  </Story>
</Canvas>


### Variant

You can use the `variant` prop to modify the forme button. This props can be declined in 2 ways, `outlined` and `text`. 

#### `variant` : outlined

<Canvas>
  <Story name="Button/Variant">
      <Button
        notice="primary"
        handleClick={() => console.log("here")}
        variant="outlined"
        children="primary"
      />
      <Button
        notice="secondary"
        handleClick={() => console.log("here")}
        variant="outlined"
        children="primary"
      />
      <Button
        handleClick={() => console.log("here")}
        variant="outlined"
        children="default"
      />
      <Button
        notice="success"
        handleClick={() => console.log("here")}
        variant="outlined"
        children="success"
      />
      <Button
        notice="info"
        handleClick={() => console.log("here")}
        variant="outlined"
        children="info"
      />
      <Button
        notice="error"
        handleClick={() => console.log("here")}
      variant="outlined"
        children="error"
      />
      <Button
        notice="warning"
        handleClick={() => console.log("here")}
        variant="outlined"
        children="warning"
      />
  </Story>
</Canvas>

#### `variant` : text

<Canvas>
  <Story name="Button/Variant-text">
    <Button
      notice="primary"
      handleClick={() => console.log("here")}
      variant="text"
      children="primary"
    />
    <Button
      notice="secondary"
      handleClick={() => console.log("here")}
      variant="text"
      children="primary"
    />
    <Button
      handleClick={() => console.log("here")}
      variant="text"
      children="default"
    />
    <Button
      notice="success"
      handleClick={() => console.log("here")}
      variant="text"
      children="success"
    />
    <Button
      notice="info"
      handleClick={() => console.log("here")}
      variant="text"
      children="info"
    />
    <Button
      notice="error"
      handleClick={() => console.log("here")}
      variant="text"
      children="error"
    />
    <Button
      notice="warning"
      handleClick={() => console.log("here")}
      variant="text"
      children="warning"
    />
  </Story>
</Canvas>

### Size

You can use the size prop to modify the button size. You can set the value to `xsmal`, `small`, `medium` or `default`. 
The default size is default and therefore doesn't need to be specified.

<Canvas>
  <Story name="Button/Size">
    <Button
        handleClick={() => console.log("here")}
        size="xsmall"
        children="Xsmall"
      />
    <Button
        handleClick={() => null}
        size="small"
        children="Small"
      />
    <Button
        handleClick={() => null}
        size="default"
        children="Default"
      />
    <Button
        handleClick={() => null}
        size="medium"
        children="Medium"
      />
  </Story>
</Canvas>

### isBlock

You can use attribute `isBlock` prop when you want the button to take the full width of its parent element.

<Canvas>
  <Story name="Button/isBlock">
    <Button
        notice="primary"
        handleClick={() => null}
        children="block Button"
        isBlock
      />
  </Story>
</Canvas>

