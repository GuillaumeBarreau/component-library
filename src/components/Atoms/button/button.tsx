import React from "react";
import classNames from "classnames";
import { ComponentButtonProps } from "./button.d";
import "./button.css";
import { Typography } from "components/Atoms/typography";
import { Icon } from "components/Atoms/icon";
import { SizeProp } from "@fortawesome/fontawesome-svg-core";

export const Button: React.FC<ComponentButtonProps> = (props) => {
  const {
    className,
    handleClick,
    children,
    iconName,
    size = "default",
    isBlock,
    variant,
    notice = "default",
  } = props;

  const iconNameSize: { [key: string]: SizeProp | "2x" } = {
    xsmall: "lg",
    small: "lg",
    medium: "2x",
    default: "2x",
  };

  const iconNode: JSX.Element | null = iconName ? (
    <Icon icon={iconName} size={`${iconNameSize[size]}`} />
  ) : !children && !iconName ? (
    <Icon icon={"bug"} size={"xs"} />
  ) : null;

  const prefixCls = "button";

  const classes = classNames(
    prefixCls,
    {
      [`${prefixCls}--notice-${notice}`]: notice,
      [`${prefixCls}--size-${size}`]: size,
      [`${prefixCls}--block`]: isBlock,
      [`${prefixCls}--variant-${variant}`]: variant,
    },
    className
  );

  const ButtonDisabled: boolean = notice === "disabled" ? true : false;

  const buttonNode = (
    <>
      {(!children && (
        <button
          className={"button--variant-icon"}
          onClick={handleClick}
          arial-label="trigger button"
        >
          <Icon icon={iconName || "bug"} size={"lg"} />
        </button>
      )) || (
        <button
          className={classes}
          onClick={handleClick}
          disabled={ButtonDisabled}
          arial-label="trigger button"
        >
          {iconNode}
          <Typography
            className="button--override-typography"
            font="bold"
            size={size}
            variant={"p"}
          >
            {children}
          </Typography>
        </button>
      )}
    </>
  );

  return buttonNode;
};
