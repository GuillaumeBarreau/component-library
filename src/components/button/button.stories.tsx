import React from 'react'
import Button from './button'
import { Meta } from '@storybook/react/types-6-0';
import { ComponentButtonProps } from './button.d';
import '../colors/classicTheme.css';

export default {
    component: Button,
    title: 'Button',
    argTypes: {
        notice: {
            control: 'select',
            options: ["default", "info", "warning", "success", "error", "disabled"],
        },
        className: {
            description: 'className content',
            control: 'text',
        },
        variant: {
            control: 'select',
            options: ["h1", "h2", "h3", "h4", "h5", "h5", "p"],
        },
        type: {
            control: 'select',
            options: ["submit"],
        },
        iconName: {
            control: 'select',
            options: [""],
        },
        children: {
            description: 'Text content',
        },
    },
    parameters: {
        a11y: {
            // optional selector which element to inspect
            element: '#root',
            // optional flag to prevent the automatic check
            manual: false,
        },
    },
} as Meta;

const Template:any = (args: JSX.IntrinsicAttributes & ComponentButtonProps) => <Button {...args} />;

export const ButtonDefault = Template.bind({})
ButtonDefault.args = {
    children: "button",
    handleClick: () => null,
}

export const ButtonSuccess = Template.bind({})
ButtonSuccess.args = {
    notice: "success",
    children: "button-success",
    handleClick: () => null,
}

export const ButtonWarning = Template.bind({})
ButtonWarning.args = {
    notice: "warning",
    children: "button-warning",
    handleClick: () => null,
}

export const ButtonError = Template.bind({})
ButtonError.args = {
    notice: "error",
    children: "button-error",
    handleClick: () => null,
}

export const ButtonInfo = Template.bind({})
ButtonInfo.args = {
    notice: "info",
    children: "button-info",
    handleClick: () => null,
}

export const ButtonDisabled = Template.bind({})
ButtonDisabled.args = {
    notice: "disabled",
    children: "button-disabled",
    handleClick: () => null,
}