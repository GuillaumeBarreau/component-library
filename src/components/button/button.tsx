import React from 'react';
import classNames from 'classnames';
import { ComponentButtonProps } from './button.d';
import "./button.css";
import Typography from "../typography";

const Button: React.FC<ComponentButtonProps> = (props) => {

  const { className, handleClick, children, iconName, type, notice = "default"} = props;
  const iconNode: string | null = iconName || null;
  const prefixCls: string  = 'button';
  
  const classes = classNames(
    prefixCls,
    {
      [`${prefixCls}-${type}`]: type,
      [`${prefixCls}-${notice}`]: notice,
    },
    className,
  );

  const buttonNode = (
    <button
      className={classes}
      onClick={handleClick}
    >
      {iconNode}
      <Typography 
        font="bold" 
        size="small"
        variant={"p"}
      >
        {children}
      </Typography>
    </button>
  );


  return buttonNode;
};

export default Button;