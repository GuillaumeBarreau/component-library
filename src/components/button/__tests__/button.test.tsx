import React, { Component } from 'react';
import { mount, shallow } from 'enzyme';
import { act } from 'react-dom/test-utils';
import Button from '../button';
import { ComponentButtonProps } from '../button.d';

describe('Button', () => {
  let defaultProps: JSX.IntrinsicAttributes & ComponentButtonProps;

  beforeEach(() => {
    defaultProps = {
      className: "",
      handleClick: jest.fn(),
      iconName: "",
      type: "",
      status: undefined,
      children: "Button"
    }
  });

  it('renders correctly', () => {
    const wrapper = shallow(
      <Button {...defaultProps}/>
    );
    expect(wrapper.debug()).toMatchSnapshot();
  });

  it('mount correctly', () => {
    const wrapper = shallow(
      <Button {...defaultProps} />
    );
    expect(() => wrapper).not.toThrow();
  });

  it('to contain defaultProps:children correctly', () => {
    const wrapper = shallow(
      <Button {...defaultProps} />
    );
    expect(() => wrapper).not.toThrow();
    expect(wrapper.html()).toContain('Button');
  });

  it('to contain defaultProps:status correctly', () => {
    const wrapper = shallow(
      <Button {...defaultProps} />
    );
    expect(() => wrapper).not.toThrow();
    expect(wrapper.html()).toContain('button-default');
  });

});