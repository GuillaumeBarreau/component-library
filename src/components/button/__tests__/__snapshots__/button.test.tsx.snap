// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button renders correctly 1`] = `
"<button className=\\"button button-default\\" onClick={[Function: mockConstructor] { _isMockFunction: true, getMockImplementation: [Function (anonymous)], mock: { calls: [], instances: [], invocationCallOrder: [], results: [] }, mockClear: [Function (anonymous)], mockReset: [Function (anonymous)], mockRestore: [Function (anonymous)], mockReturnValueOnce: [Function (anonymous)], mockResolvedValueOnce: [Function (anonymous)], mockRejectedValueOnce: [Function (anonymous)], mockReturnValue: [Function (anonymous)], mockResolvedValue: [Function (anonymous)], mockRejectedValue: [Function (anonymous)], mockImplementationOnce: [Function (anonymous)], mockImplementation: [Function (anonymous)], mockReturnThis: [Function (anonymous)], mockName: [Function (anonymous)], getMockName: [Function (anonymous)] }}>
  <Typography font=\\"bold\\" size=\\"small\\">
    Button
  </Typography>
</button>"
`;
